mkosi/distributions/rhel_ubi.py

```diff
@@ -0,0 +1,153 @@
+            "zstd",
+        ]
+
+    @classmethod
+    def setup(cls, state: MkosiState) -> None:
+        release = int(state.config.release)
+
+        if release <= 7:
+            die("UBI-7 or earlier variants are not supported")
+
+        setup_dnf(state, cls.repositories(state.config, release))
+
+    @classmethod
+    def install(cls, state: MkosiState) -> None:
+        # Make sure glibc-minimal-langpack is installed instead of glibc-all-langpacks.
+        cls.install_packages(state, ["filesystem", "glibc-minimal-langpack"], apivfs=False)
+
+        # On Fedora, the default rpmdb has moved to /usr/lib/sysimage/rpm so if that's the case we
+        # need to move it back to /var/lib/rpm.
+        move_rpm_db(state.root)
+
+    @classmethod
+    def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bool = True) -> None:
+        invoke_dnf(state, "install", packages, apivfs=apivfs)
+
+    @classmethod
+    def remove_packages(cls, state: MkosiState, packages: Sequence[str]) -> None:
+        invoke_dnf(state, "remove", packages)
+
+    @staticmethod
+    def architecture(arch: Architecture) -> str:
+        a = {
+            Architecture.x86_64   : "x86_64",
+            Architecture.ppc64_le : "ppc64le",
+            Architecture.s390x    : "s390x",
+            Architecture.arm64    : "aarch64",
+        }.get(arch)
+
+        if not a:
+            die(f"Architecture {a} is not supported by UBI")
+
+        return a
+
+    @staticmethod
+    def gpgurls() -> tuple[str, ...]:
+        return (
+            "https://access.redhat.com/security/data/fd431d51.txt",
+        )
+
+    @classmethod
+    def _repository_variants(cls, config: MkosiConfig, repo: str) -> list[Repo]:
+        if config.local_mirror:
+            yield Repo(repo, f"baseurl={config.local_mirror}", cls.gpgurls())
+        else:
+            v = config.release
+            yield Repo(
+                f"ubi-{v}-{repo}-rpms",
+                f"baseurl={join_mirror(config, f'ubi{v}/{v}/$basearch/{repo}/os')}",
+                cls.gpgurls(),
+            )
+            yield Repo(
+                f"ubi-{v}-{repo}-debug-rpms",
+                f"baseurl={join_mirror(config, f'ubi{v}/{v}/$basearch/{repo}/debug')}",
+                cls.gpgurls(),
+                enabled=False,
+            )
+            yield Repo(
+                f"ubi-{v}-{repo}-source",
+                f"baseurl={join_mirror(config, f'ubi{v}/{v}/$basearch/{repo}/source')}",
+                cls.gpgurls(),
+                enabled=False,
+            )
+            if repo == "codeready-builder":
+                yield Repo(
+                    f"ubi-{v}-{repo}",
+                    f"baseurl={join_mirror(config, f'ubi{v}/{v}/$basearch/{repo}/os')}",
+                    cls.gpgurls(),
+                    enabled=False,
+                )
+
+    @classmethod
+    def repositories(cls, config: MkosiConfig, release: int) -> list[Repo]:
+        repos = [
```

Comment:
I opted to use a generator because then we can deal with the conditional `codeready-builder` part nicely. We just have an extra `yield`, no need to concatenate lists.